src/main/java/net/openhft/chronicle/queue/internal/reader/queueentryreaders/VanillaQueueEntryReader.java

```diff
@@ -0,0 +1,52 @@
+package net.openhft.chronicle.queue.internal.reader.queueentryreaders;
+
+import net.openhft.chronicle.queue.ExcerptTailer;
+import net.openhft.chronicle.queue.reader.MessageConsumer;
+import net.openhft.chronicle.queue.reader.QueueEntryHandler;
+import net.openhft.chronicle.queue.reader.QueueEntryReader;
+import net.openhft.chronicle.queue.reader.ReadResult;
+import net.openhft.chronicle.wire.DocumentContext;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+public final class VanillaQueueEntryReader implements QueueEntryReader {
+
+    private final ExcerptTailer tailer;
+    private final Function<ExcerptTailer, DocumentContext> pollMethod;
+    private final QueueEntryHandler messageConverter;
+    private final MessageConsumer messageConsumer;
+    private final ThreadLocal<AtomicInteger> tlMatchCounter = ThreadLocal.withInitial(AtomicInteger::new);
+
+    public VanillaQueueEntryReader(@NotNull ExcerptTailer tailer, @NotNull Function<ExcerptTailer, DocumentContext> pollMethod,
+                                   @NotNull QueueEntryHandler messageConverter, @NotNull MessageConsumer messageConsumer) {
+        this.tailer = tailer;
+        this.pollMethod = pollMethod;
+        this.messageConverter = messageConverter;
+        this.messageConsumer = messageConsumer;
+    }
+
+    @Override
+    public ReadResult read() {
+        try (DocumentContext dc = pollMethod.apply(tailer)) {
+            if (!dc.isPresent()) {
+                return ReadResult.NOT_FOUND;
+            }
+
+            final int matchCount = runMessageConverter(dc);
+            return new ReadResult(true, matchCount);
+        }
+    }
+
+    private int runMessageConverter(DocumentContext documentContext) {
+        AtomicInteger matchCounter = tlMatchCounter.get();
+        matchCounter.set(0);
+        messageConverter.accept(documentContext.wire(), val -> {
+            if (messageConsumer.consume(documentContext.index(), val)) {
+                matchCounter.incrementAndGet();
+            }
+        });
```

Comment:
This is a bit boiler-platey due to the mismatch between what the `messageConverter` expects to write to and the new MessageConsumer interface.